File: backend/module1_learner/extractors/causal_chains.py
#!/usr/bin/env python3
"""
Causal Chain Extractor
Discovers event sequences and cause-effect relationships
"""

from collections import defaultdict

class CausalChainExtractor:
    def __init__(self):
        self.chains = []
        self.sequence_probabilities = {}
        
    def extract(self, logs):
        """Extract causal chains from logs"""
        print("   - Building event sequences...")
        
        # Group by component and user to find chains
        chains = self._find_sequences_by_component(logs)
        
        # Calculate transition probabilities
        probabilities = self._calculate_transition_probabilities(chains)
        
        # Find common patterns
        common_patterns = self._find_common_patterns(chains)
        
        return {
            'chains': chains[:20],  # Top 20 chains
            'transition_probabilities': probabilities,
            'common_patterns': common_patterns,
            'total_chains': len(chains)
        }
    
    def _find_sequences_by_component(self, logs):
        """Group logs by component and find sequences"""
        component_logs = defaultdict(list)
        
        # Group by component
        for log in logs:
            if 'component' in log:
                component_logs[log['component']].append(log)
        
        # Find sequences within each component
        sequences = []
        for component, logs_list in component_logs.items():
            if len(logs_list) > 1:
                for i in range(len(logs_list) - 1):
                    current = logs_list[i]['message'][:50] if 'message' in logs_list[i] else ''
                    next_msg = logs_list[i+1]['message'][:50] if 'message' in logs_list[i+1] else ''
                    
                    sequences.append({
                        'component': component,
                        'from': current,
                        'to': next_msg,
                        'time_diff': self._get_time_diff(logs_list[i], logs_list[i+1])
                    })
        
        return sequences
    
    def _get_time_diff(self, log1, log2):
        """Calculate time difference between logs"""
        # Simplified - would need proper timestamp parsing
        return 1  # Placeholder
    
    def _calculate_transition_probabilities(self, sequences):
        """Calculate probability of event B following event A"""
        transitions = defaultdict(lambda: defaultdict(int))
        totals = defaultdict(int)
        
        for seq in sequences:
            from_event = seq['from']
            to_event = seq['to']
            
            transitions[from_event][to_event] += 1
            totals[from_event] += 1
        
        # Convert to probabilities
        probabilities = {}
        for from_event, to_events in transitions.items():
            probabilities[from_event] = {
                to_event: count / totals[from_event]
                for to_event, count in to_events.items()
            }
        
        return dict(sorted(probabilities.items(), 
                          key=lambda x: sum(x[1].values()), reverse=True)[:20])
    
    def _find_common_patterns(self, sequences):
        """Find common 3-event patterns"""
        patterns = defaultdict(int)
        
        for i in range(len(sequences) - 2):
            pattern = (
                sequences[i]['from'],
                sequences[i+1]['from'],
                sequences[i+2]['from']
            )
            patterns[pattern] += 1
        
        common = []
        for pattern, count in sorted(patterns.items(), key=lambda x: x[1], reverse=True)[:10]:
            common.append({
                'pattern': list(pattern),
                'frequency': count
            })
        
        return common

File: log-collector/internal/learner/causal_chains.go
package learner

import (
	"sort"
	"strings"

	"migrate/internal/models"
)

// CausalChainExtractor extracts event sequences and cause-effect relationships
type CausalChainExtractor struct{}

// NewCausalChainExtractor creates a new causal chain extractor
func NewCausalChainExtractor() *CausalChainExtractor {
	return &CausalChainExtractor{}
}

// Extract extracts causal chains
func (cce *CausalChainExtractor) Extract(logs []models.LogEntry) models.CausalChains {
	chains := cce.findSequencesByComponent(logs)
	probabilities := cce.calculateTransitionProbabilities(chains)
	patterns := cce.findCommonPatterns(chains)
	
	return models.CausalChains{
		Chains:                 cce.topChains(chains, 20),
		TransitionProbabilities: probabilities,
		CommonPatterns:         patterns,
		TotalChains:            len(chains),
	}
}

// ChainData represents a sequence event
type ChainData struct {
	Component string
	From      string
	To        string
	TimeDiff  int
}

// findSequencesByComponent groups logs by component and finds sequences
func (cce *CausalChainExtractor) findSequencesByComponent(logs []models.LogEntry) []ChainData {
	// Group by component
	componentLogs := make(map[string][]models.LogEntry)
	for _, log := range logs {
		component, ok := log.Fields["component"].(string)
		if !ok {
			component = "unknown"
		}
		componentLogs[component] = append(componentLogs[component], log)
	}
	
	sequences := make([]ChainData, 0)
	
	for _, logsList := range componentLogs {
		if len(logsList) < 2 {
			continue
		}
		
		for i := 0; i < len(logsList)-1; i++ {
			currentMsg, ok1 := logsList[i].Fields["message"].(string)
			nextMsg, ok2 := logsList[i+1].Fields["message"].(string)
			
			if !ok1 || !ok2 {
				continue
			}
			
			// Use first 50 chars as event identifier
			if len(currentMsg) > 50 {
				currentMsg = currentMsg[:50]
			}
			if len(nextMsg) > 50 {
				nextMsg = nextMsg[:50]
			}
			
			component, _ := logsList[i].Fields["component"].(string)
			
			timeDiff := 1
			if !logsList[i].Timestamp.IsZero() && !logsList[i+1].Timestamp.IsZero() {
				timeDiff = int(logsList[i+1].Timestamp.Sub(logsList[i].Timestamp).Seconds())
			}
			
			sequences = append(sequences, ChainData{
				Component: component,
				From:      currentMsg,
				To:        nextMsg,
				TimeDiff:  timeDiff,
			})
		}
	}
	
	return sequences
}

// calculateTransitionProbabilities calculates probability of B following A
func (cce *CausalChainExtractor) calculateTransitionProbabilities(sequences []ChainData) map[string]map[string]float64 {
	transitions := make(map[string]map[string]int)
	totals := make(map[string]int)
	
	for _, seq := range sequences {
		if _, exists := transitions[seq.From]; !exists {
			transitions[seq.From] = make(map[string]int)
		}
		transitions[seq.From][seq.To]++
		totals[seq.From]++
	}
	
	// Convert to probabilities
	probabilities := make(map[string]map[string]float64)
	for from, toMap := range transitions {
		probabilities[from] = make(map[string]float64)
		for to, count := range toMap {
			if totals[from] > 0 {
				probabilities[from][to] = float64(count) / float64(totals[from])
			}
		}
	}
	
	// Sort and return top 20 by total frequency
	type fromTotal struct {
		from  string
		total int
	}
	
	var sorted []fromTotal
	for from, total := range totals {
		sorted = append(sorted, fromTotal{from, total})
	}
	
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].total > sorted[j].total
	})
	
	result := make(map[string]map[string]float64)
	for i, ft := range sorted {
		if i >= 20 {
			break
		}
		result[ft.from] = probabilities[ft.from]
	}
	
	return result
}

// findCommonPatterns finds common 3-event patterns
func (cce *CausalChainExtractor) findCommonPatterns(sequences []ChainData) []models.Pattern {
	if len(sequences) < 3 {
		return []models.Pattern{}
	}
	
	patternCounts := make(map[string]int)
	
	for i := 0; i < len(sequences)-2; i++ {
		pattern := sequences[i].From + "|" + sequences[i+1].From + "|" + sequences[i+2].From
		patternCounts[pattern]++
	}
	
	// Sort patterns
	type patternCount struct {
		pattern string
		count   int
	}
	
	var sorted []patternCount
	for p, c := range patternCounts {
		sorted = append(sorted, patternCount{p, c})
	}
	
	sort.Slice(sorted, func(i, j int) bool {
		return sorted[i].count > sorted[j].count
	})
	
	// Convert to Pattern structs
	patterns := make([]models.Pattern, 0)
	for i, pc := range sorted {
		if i >= 10 {
			break
		}
		
		parts := strings.Split(pc.pattern, "|")
		patterns = append(patterns, models.Pattern{
			Pattern:   parts,
			Frequency: pc.count,
		})
	}
	
	return patterns
}

// topChains returns the top N chains
func (cce *CausalChainExtractor) topChains(sequences []ChainData, n int) []models.Chain {
	if len(sequences) <= n {
		chains := make([]models.Chain, len(sequences))
		for i, seq := range sequences {
			chains[i] = models.Chain{
				Component: seq.Component,
				From:      seq.From,
				To:        seq.To,
				TimeDiff:  seq.TimeDiff,
			}
		}
		return chains
	}
	
	chains := make([]models.Chain, n)
	for i := 0; i < n; i++ {
		chains[i] = models.Chain{
			Component: sequences[i].Component,
			From:      sequences[i].From,
			To:        sequences[i].To,
			TimeDiff:  sequences[i].TimeDiff,
		}
	}
	return chains
}
